pkg/cli/admin/release/info.go

```diff
@@ -1987,6 +2035,306 @@ func describeBugs(out, errOut io.Writer, diff *ReleaseDiff, dir string, format s
 	return nil
 }
 
+func (o *InfoOptions) listRpmdb(releaseInfo *ReleaseInfo, dir, format, targetImage string) error {
+	if targetImage == "" {
+		// find image for machine-os component
+		if tag, ok := releaseInfo.ComponentTags["machine-os"]; !ok {
+			return fmt.Errorf("component machine-os not found in release")
+		} else {
+			targetImage = tag
+		}
+	}
+
+	var err error
+	var targetTag *imageapi.TagReference
+	for _, tag := range releaseInfo.References.Spec.Tags {
+		if tag.Name == targetImage {
+			targetTag = &tag
+			break
+		}
+	}
+	if targetTag == nil {
+		return fmt.Errorf("tag %s not found in release", targetImage)
+	}
+
+	ref, err := imagereference.Parse(targetTag.From.Name)
+	if err != nil {
+		return err
+	}
+
+	// not sure this can ever happen, but just sanity-check anyway
+	if ref.ID == "" {
+		return fmt.Errorf("image not digest-based: %s", ref)
+	}
+
+	dbpath := filepath.Join(dir, ref.ID)
+	if _, err = os.Stat(dbpath); os.IsNotExist(err) {
+		err = o.downloadRpmdb(ref, dbpath)
+	}
+	if err != nil {
+		return err
+	}
+
+	db, err := loadRpmdb(dbpath)
+	if err != nil {
+		return err
+	}
+
+	if format == "json" {
+		data, err := json.MarshalIndent(db, "", "  ")
+		if err != nil {
+			return err
+		}
+		fmt.Fprintln(o.Out, string(data))
+	} else {
+		pkgs := make([]string, 0, len(db))
+		for pkg := range db {
+			pkgs = append(pkgs, pkg)
+		}
+		sort.Strings(pkgs)
+		fmt.Fprintf(o.Out, "Package contents:\n")
+		for _, pkg := range pkgs {
+			fmt.Fprintf(o.Out, "  %s-%s\n", pkg, db[pkg])
+		}
+	}
+
+	return nil
+}
+
+func (o *InfoOptions) describeRpmDiff(releaseInfo *ReleaseInfo, diff *ReleaseDiff, dir, format, targetImage string) error {
+	if diff.To.Digest == diff.From.Digest {
+		return fmt.Errorf("releases are identical")
+	}
+
+	if targetImage == "" {
+		// find image for machine-os component
+		if tag, ok := releaseInfo.ComponentTags["machine-os"]; !ok {
+			return fmt.Errorf("component machine-os not found in release")
+		} else {
+			targetImage = tag
+		}
+	}
+
+	osDiff, ok := diff.ChangedImages[targetImage]
+	if !ok {
+		// image didn't change; we're done!
+		if format == "json" {
+			fmt.Println("{}")
+		}
+		return nil
+	}
+
+	var err error
+	var fromRef, toRef imagereference.DockerImageReference
+	if fromRef, err = imagereference.Parse(osDiff.From.From.Name); err != nil {
+		return err
+	}
+	if toRef, err = imagereference.Parse(osDiff.To.From.Name); err != nil {
+		return err
+	}
+
+	// not sure this can ever happen, but just sanity-check anyway
+	if fromRef.ID == "" {
+		return fmt.Errorf("image not digest-based: %s", fromRef)
+	}
+	if toRef.ID == "" {
+		return fmt.Errorf("image not digest-based: %s", toRef)
```

Comment:
nit: to make debugging slightly easier, can we include some distinguishing string in these context wrappers?  `initial machine-os image not digest-based` and `new machine-os image not digest-based` or some such?  There's an earlier `image not digest-based` error around line 2067 that could also be uniquified somehow.  I'm just trying to set things up so that if the command exits with these errors, the way the offending ref was extracted from the input arguments is somewhat clear, and which of these several image-lookup paths lead to the surprising data.

Reply:
Yes, I _usually_ do this too, for the same reason. :)